src/Corax/IndexEntry/IndexEntryWriter.cs

```diff
@@ -566,33 +634,27 @@ public ByteStringContext<ByteStringMemoryCache>.InternalScope Finish(out ByteStr
 
         // The size of the known fields metadata section
         int metadataSection = IndexEntryReader.TableEncodingLookupTable[(int)encodeSize] * _knownFields.Count;
-        // The size of the unknown/dynamic fields metadata section            
-        int dynamicMetadataSection = _dynamicFieldIndex * sizeof(uint);
-        int dynamicMetadataSectionOffset = buffer.Length - dynamicMetadataSection - 1;
 
         ref var header = ref MemoryMarshal.AsRef<IndexEntryHeader>(buffer);
-        header.Length = (uint)(_dataIndex + dynamicMetadataSection + metadataSection + 1);
 
         // The known field count is encoded as xxxxxxyy where:
-        // x: the count
+        // x: the count (1 ... 16K should be enough, using checked math to verify it anyway)
         // y: the encode size
-        header.KnownFieldCount = (ushort)(_knownFields.Count << 2 | (int)encodeSize);
-        header.DynamicTable = (uint)_dataIndex;
+        header.KnownFieldCount = checked((ushort)(_knownFields.Count << 2 | (int)encodeSize));
 
         // The dynamic metadata fields count. 
-        Unsafe.WriteUnaligned(ref buffer[_dataIndex], (byte)_dynamicFieldIndex);
-        _dataIndex += sizeof(byte);
-
-        if (dynamicMetadataSection != 0)
+        header.DynamicTable = (uint)_dataIndex;
+        _dataIndex += VariableSizeEncoding.Write(buffer, numberOfDynamicFields, _dataIndex);
+        if (_dynamicFieldsLocations != null)
         {
-            // From the offset to the end... move the data toward the closest position
-            var metadataTable = buffer[dynamicMetadataSectionOffset..];
-            metadataTable.CopyTo(buffer[_dataIndex..]);
-
-            // Move the pointer to the end of the copied section.
-            _dataIndex += dynamicMetadataSection;
+            foreach (int fieldsLocation in _dynamicFieldsLocations!)
+            {
+                _dataIndex += VariableSizeEncoding.Write(buffer, fieldsLocation, _dataIndex);
+            }
```

Comment:
This should use the same scheme that we use for writing the known metadata table. Where depending on the size it will use fixed size values. This saves a lot of time when searching for an specific dynamic name.

Reply:
Why? We are not going to be doing actual searches here. And the expected scenario is to have very small number of dynamic fields.
And _searching_ for them is even more rare. Furthermore, the setup is actually more complex and space consuming than varints in most cases.

Reply:
As long as we are ok to hydrate it in a temporary, it can be done in that way.

Reply:
See how this is handled here, we are iterating over the values.

https://github.com/ravendb/ravendb/pull/14915/files#diff-616ac881847af034e4b7eda9043eab32ccfa0f003d7cac89a27a443fde38ddeeR483-R493